src/catchup/CatchupManagerImpl.cpp

```diff
@@ -93,4 +200,110 @@ CatchupManagerImpl::logAndUpdateCatchupStatus(bool contiguous)
 {
     logAndUpdateCatchupStatus(contiguous, getStatus());
 }
+
+bool
+CatchupManagerImpl::hasBufferedLedger() const
+{
+    return !mSyncingLedgers.empty();
+}
+
+LedgerCloseData const&
+CatchupManagerImpl::getBufferedLedger() const
+{
+    if (!hasBufferedLedger())
+    {
+        throw std::runtime_error(
+            "getBufferedLedger called when mSyncingLedgers is empty!");
+    }
+
+    return mSyncingLedgers.front();
+}
+
+void
+CatchupManagerImpl::popBufferedLedger()
+{
+    if (!hasBufferedLedger())
+    {
+        throw std::runtime_error(
+            "popBufferedLedger called when mSyncingLedgers is empty!");
+    }
+
+    mSyncingLedgers.pop_front();
+}
+
+void
+CatchupManagerImpl::syncMetrics()
+{
+    mSyncingLedgersSize.set_count(mSyncingLedgers.size());
+}
+
+void
+CatchupManagerImpl::trimAndReset()
+{
+    assert(mCatchupWork);
+
+    if (mCatchupWork->getState() == BasicWork::State::WORK_SUCCESS)
+    {
+        assert(mSyncingLedgers.empty());
```

Comment:
Why this assert? This may be true (hard to tell), and it makes assumptions on what ledger manager is doing and how `mCatchupWork` is stopping.

Reply:
Well `mCatchupWork` can only succeed if it applies all buffered ledgers, so I think that assumption makes sense. `trimAndReset` is also called at the top of this method before buffered ledgers are added, so I don't see why ledger manager matters in this case.

Reply:
I am talking about the transition from `mCatchupWork` almost done to done: depending on how `mCatchupWork` schedules work, it will still show up as `!done()` so ledger manager will continue to call `processLedger` even though they may not get picked up

Reply:
actually, I did the analysis and right now it looks like the way `Work` works is that it always calls the main `doWork` (so `runCatchupStep`) last *and* marks the work done in one `crankWork` invocation, so it's probably OK

Reply:
Moved the assert to `CatchupWork` like we talked about.